Implementation plan: Catalog Data Cleaning (#3848)
WordPress/openverse: documentation/projects/proposals/data_normalization/20240227-implementation_plan_catalog_data_cleaning.md

# 2024-02-27 Implementation Plan: Catalog Data Cleaning

**Author**: @krysal

## Reviewers

- [ ] @obulat
- [ ] @AetherUnbound

## Project links

<!-- Enumerate any references to other documents/pages, including milestones and other plans -->

- [Project Thread](https://github.com/WordPress/openverse/issues/430)

This project does not have a project proposal because the scope and rationale of
the project are clear, as defined in the project thread. In doubt, check the
[Expected Outcomes](#expected-outcomes) section below.

## Overview

> This document describes a solution for incorrect data in the catalog database
> (DB) that has to be cleaned up every time a data refresh is run, avoiding wasted
> resources.

Reviewer: Nit for rewording:
```suggestion
This document describes a mechanism for rectifying incorrect data in the catalog database
(DB) that currently has to be cleaned up every time a data refresh is run. This one-time
fix is an effort to avoid wasting resources and data refresh runtime.
```

## Background

One of the steps of the [data refresh process for images][img-data-refresh] is
cleaning the data that is not fit for production. This process is triggered

> weekly by an Airflow DAG, and then runs in the Ingestion Server, taking

Reviewer: ```suggestion
weekly by an Airflow DAG, which then runs in the Ingestion Server, taking
```

approximately just over **20 hours** to complete, according to a inspection of

> latest executions. The cleaned data is only saved to the API database, which is

Reviewer: Nit:

```suggestion
recent executions as of the time of drafting this document. The cleaned data is only saved to the API database, which is
```

> replaced each time during the same data refresh, causing it to have to be
> repeated each time to make the _same_ corrections.

Reviewer: Nit:
```suggestion
replaced each time during the same data refresh, meaning this process has to be
repeated each time to make the _same_ corrections.
```

> This cleaning process was designed this way to speed the rows update up since
> the relevant part was to provide the correct data to users via the API. Most of

Reviewer: Phrasing:
```suggestion
This cleaning process was designed this way to optimize writes to the catalog, since
the most important factor was to provide the correct data to users via the API. Most of
```

Reviewer: I updated the suggestion to use "upstream database" instead, does it look okay now?

Reviewer 2: I agree that this sentence needs clarification since it's not clear if we mean "catalog db rows update" or the "API db rows update".

Author: I see! In this case, it should be the API DB then, because the Ingestion Serves saves the cleaned data there.

Applied 👍

> the rows affected were added previous to the creation of the `MediaStore` class

Reviewer: ```suggestion
the rows affected were added prior to the creation of the `MediaStore` class
```

in the Catalog (possibly by the discontinued CommonCrawl ingestion) which is

> nowadays responsible for validating the provider data. However, it entails a

Reviewer: Nit (just to contrast with the previous method!):
```suggestion
nowadays responsible for validating the provider data prior to upserting the records into the Catalog. However, the current approach entails a
```

Author: Same [issue](https://github.com/WordPress/openverse/pull/3848#discussion_r1511575727) as before, I think.

Reviewer: Also updated to use "upstream database" instead!

problem of wasting resources both in time, which continues to increase, and in
the machines (CPU) it uses, which could easily be avoided making the changes
permanent by saving them in the upstream database.

[img-data-refresh]:

>   https://github.com/WordPress/openverse-catalog/blob/main/DAGs.md#image_data_refresh

Reviewer: This link 404s for me, and looks to be pointing to our archived catalog repo.

Author: Ups, fixed!

## Expected Outcomes

<!-- List any succinct expected products from this implementation plan. -->

> - The catalog database (upstream) preserves the cleaned data results of the
>   current Ingestion Server's cleaning steps

Reviewer: Nit:
```suggestion
- The catalog database (upstream) contains the cleaned data outputs of the
  current Ingestion Server's cleaning steps
```

- The image Data Refresh process is simplified by reducing the cleaning steps

>   time to nearly zero (and optionally removing them).

Reviewer: The removal of the cleaning steps from the ingestion server should be part of the success criteria for this IP, IMO!

Author: Changed for it 🔥

## Step-by-step plan

The cleaning functions that the Ingestion Server applies are already implemented

> in the Catalog in the `MediaStore` class: see its `_tag_blacklisted` method

Reviewer: Do you mind linking to the relevant code here? And agreed on a rename 😄

Author: Done ✅

(which probably should be renamed) and the [url utilities][url_utils] file. The
only part that it's not there and can't be ported is the filtering of
low-confidence tags, since provider scripts don't save an "accuracy" by tag.

With this the plan then starts in the Ingestion Server with the following steps:

1. [Save TSV files of cleaned data to AWS S3](#save-tsv-files-of-cleaned-data-to-aws-s3)
1. [Make and run a batched update DAG for one-time cleanup](#make-and-run-a-batched-update-dag-for-one-time-cleanup)
1. [Run an image Data Refresh to confirm cleaning time is reduced](#run-an-image-data-refresh-to-confirm-cleaning-time-is-reduced)

[url_utils]:
  https://github.com/WordPress/openverse/blob/a930ee0f1f116bac77cf56d1fb0923989613df6d/catalog/dags/common/urls.py

## Step details

### Save TSV files of cleaned data to AWS S3

> In a previous exploration, it was set to store TSV files of the cleaned data in

Reviewer: ```suggestion
In a previous exploration, the Ingestion Server was set to store TSV files of the cleaned data in
```

Do you mind also linking the PR where we added this for context?

Author: Done ✅

the form of `<identifier> <cleaned_field>`, which can be used later to perform
the updates efficiently in the catalog DB, which only had indexes for the
`identifier` field. These files are saved to the disk of the Ingestion Server
EC2 instances, and worked fine for files with URL corrections since this type of
fields is relatively short, but became a problem when trying to save tags, as
the file turned too large and filled up the disk, causing problems to the data

> refresh execution.

Reviewer 2: Noting that the files must have a timestamp (or some indication of the date saved) in their names to make sure that we don't overwrite them on the next refresh.

Author: It doesn't matter much if the data is overwritten I think 🤔 In theory, it will always be the same given the new data is ingested cleaned.

The alternative is to upload TSV files to the Amazon Simple Storage Service
(S3), creating a new bucket or using `openverse-catalog` with a subfolder. The
benefit of using S3 buckets is that they have streaming capabilities and will
allow us to read the files in chunks later if necessary for performance. The
downside is that objects in S3 don't allow appending, so it may require to
upload files with different part numbers or evaluate if the [multipart upload

> process][aws_mpu] will serve us here.

Reviewer: Noting here that I've used the [`smart_open` library](https://github.com/piskvorky/smart_open) in the past to seamlessly "stream" larger files into S3 without having to write them to disk locally!

Author: I came across this library too but thought it was only for fetching files. It will be great if it allows us to upload the file directly to ASW! Thanks for mentioning it 😄 I'll check it out.

[aws_mpu]:
  https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html

| timestamp (UTC)     | 'url' | 'creator_url' | 'foreign_landing_url' | 'tags' |
| ------------------- | :---: | :-----------: | :-------------------: | :----: |
| 2024-02-27 04:05:26 | 22156 |    9035458    |        8809213        |   0    |
| 2024-02-20 04:06:56 | 22157 |    9035456    |        8809209        |   0    |
| 2024-02-13 04:41:22 | 22155 |    9035451    |        8809204        |   0    |

> To have some numbers of the problem we are delaing with, the previous table

Reviewer 2: ```suggestion
To have some numbers of the problem we are dealing with, the previous table
```

shows the number of records cleaned by field for last runs at the moment of
writing this IP, except for tags, which we don't have accurate registries since

> file saving was disabled.

Reviewer: Do you mind moving this section above the table for better context?

Author: I moved it one paragraph higher. Is this ok?

Reviewer: Ah, I meant having it above the table rather than below it! I thought the order of the paragraphs themselves was fine 😄

Author: Ah, I see! Now I understand it better 😆 Moved the explanatory paragraphs before the table!

### Make and run a batched update DAG for one-time cleanup

A batched catalog cleaner DAG (or potentially a `batched_update_from_file`)
should take the files of the previous step to perform an batched update on the
catalog's image table, while handling deadlocking and timeout concerns, similar
to the [batched_update][batched_update]. This table is constantly in use by

> other DAGs, such as those from API providers or the data refresh process, and

Reviewer: Nit:
```suggestion
other DAGs, such as those from providers ingestion or the data refresh process, and
```

ideally can't be singly blocked by any DAG.

[batched_update]: ./../../../catalog/reference/DAGs.md#batched_update

A [proof of concept PR](https://github.com/WordPress/openverse/pull/3601)
consisted of uploading each file to temporary `UNLOGGED` DB tables (which
provides huge gains in writing performance while their disadventages are not
relevant since they won't be permanent), and including a `row_id` serial number
used later to query it in batches. Adding an index in this last column after
filling up the table could improve the query performance. An adaptation will be

> needed to handle the column type of tags (`jsonb`).

Reviewer: Do you mind adding a line at the end here explicitly noting that we'll need to run the DAGs on the previously generated TSVs as part of this step?

Author: It's mentioned in the first line of this section 😶 

> A batched catalog cleaner DAG (or potentially a `batched_update_from_file`) **should take the files of the previous step** to...

Does it need to be rephrased or moved?

### Run an image data refresh to confirm cleaning time is reduced

Finally, after the previous steps are done, running a data refresh will confirm
there are no more updates applied at ingestion. If time isn't significantly
reduced then it will be necessary to check what was missing in the previous
steps.

If confirmed the time is reduced to zero, optionally the cleaning steps can be
removed, or leave them in case we want to perform a similar cleaning effort

> later.

Reviewer: I think we can maybe leave a cleaning "shim" in for the future, but the current cleaning logic that gets called should be removed.

Author: How will look a cleaning shim in your opinion?

Reviewer: I was thinking we could perhaps leave this section in, even if `clean_image_data` is a noop:

https://github.com/WordPress/openverse/blob/9b4f727d87e026aa2ca7fb9afcda0c05666ef9e5/ingestion_server/ingestion_server/ingest.py#L326-L333

Now that I say it though, it seems silly 😅 we probably don't need to keep that either!

Reviewer 2: I kind of think we do need to have this shim, because I'm pretty sure we will find something to clean, at least in the short run. Would be nice to have some criteria to determine when to remove the shim. Maybe, when the second stage of the data normalization project is done?

With this shim in place, the functions that clean tags and URLs in the `cleanup.py` would be removed, and this dictionary would have empty values:

https://github.com/WordPress/openverse/blob/9b4f727d87e026aa2ca7fb9afcda0c05666ef9e5/ingestion_server/ingestion_server/cleanup.py#L151-L167

Author: I was debating between the two options, too.

> Would be nice to have some criteria to determine when to remove the shim. Maybe, when the second stage of the data normalization project is done?

That's a great idea! I agree with it 💯

## Dependencies

### Infrastructure

No changes needed. The Ingestion Server already has the credentials required to
[connect with AWS](https://github.com/WordPress/openverse/blob/a930ee0f1f116bac77cf56d1fb0923989613df6d/ingestion_server/ingestion_server/indexer_worker.py#L23-L28).

<!--
### Tools & packages

 Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->

### Other projects or work

Once the steps have been completed and proved the method works we could make
additional similar corrections following the same procedure. Some potentially
related issues are:

- [Some images have duplicate incorrectly decoded unicode tags #1303](https://github.com/WordPress/openverse/issues/1303)
- [Provider scripts may include html tags in record titles #1441](https://github.com/WordPress/openverse/issues/1441)

> - [Fix Wikimedia image titles #1728](https://github.com/WordPress/openverse/issues/1728)

Reviewer: Thanks for noting these!

Reviewer 2: Another thing we could add here is `license_url` check: save all of the items with their computed `license_url` values if this URL is not set in metadata. I suspect that there might be none such items, in which case we can simplify this part of API:

https://github.com/WordPress/openverse/blob/3c5cdd069bc0a9e2aa1f979bb99a7ee1402f8560/api/api/models/media.py#L86-L92

Here's a related issue: https://github.com/WordPress/openverse/issues/552
Note that there are also probably some items with incorrectly set license pairs with metadata set to `null` that I didn't finish fixing in https://github.com/WordPress/openverse-catalog/pull/1045

Reviewer 2: One more ingestion server step we can do is remove the extraction of the extension from the image URL: https://github.com/WordPress/openverse/issues/1560

These two things are not part of the cleanup process per se, so we might want to add them to the cleanup once, specifically for saving the fixed information, and remove these steps (and the code in the ES models` as soon as we confirm that all data has been cleaned.

Author: @obulat That's a good observation on the `license_url`! I'll run a query to see if there are still items without it.

I'll add the file type/extension issue to the list ➕

Reviewer 2: Thank you for investigating, @krysal .

For the item you mention, it doesn't have `raw_license_url`, but it does have `license_url`, so it's fine. However, if you do `SELECT * FROM image WHERE meta_data->>'license_url' is null`, you get a lot of results.

I think we could run DAG from [Add a DAG for backfilling license_url when meta_data is null](https://github.com/WordPress/openverse-catalog/pull/1005) for items where `meta_data.license_url` is null. And only resort to the process from this PR if that DAG takes too long.

# 2024-02-27 Implementation Plan: Catalog Data Cleaning

**Author**: @krysal

## Reviewers

> - [ ] @obulat

Reviewer 2: ```suggestion
- [x] @obulat
```

> - [ ] @AetherUnbound

Reviewer: ```suggestion
- [X] @AetherUnbound
```

### Make and run a batched update DAG for one-time cleanups

A batched catalog cleaner DAG (or potentially a `batched_update_from_file`)
should take the files of the previous step to perform a batched update on the
catalog's image table, while handling deadlocking and timeout concerns, similar
to the [batched_update][batched_update]. This table is constantly in use by
other DAGs, such as those from providers ingestion or the data refresh process,
and ideally can't be singly blocked by any DAG.

[batched_update]: ./../../../catalog/reference/DAGs.md#batched_update

A [proof of concept PR](https://github.com/WordPress/openverse/pull/3601)
consisted of uploading each file to temporary `UNLOGGED` DB tables (which
provides huge gains in writing performance while their disadventages are not
relevant to us, they won't be permanent), and include a `row_id` serial number
used later to query it in batches. The following must be included:

- Add an index for the `identifier` column in the temporary table after filling
  it up, to improve the query performance
- An adaptation to handle the column type of tags (`jsonb`) and modify the
  `metadata`

> - Include an DAG task for reporting the number of rows affected by column to

Reviewer 2: ```suggestion
- Include a DAG task for reporting the number of rows affected by column to
```

| 2024-02-20 04:06:56 | 22157 |    9035456    |        8809209        |   0    |
| 2024-02-13 04:41:22 | 22155 |    9035451    |        8809204        |   0    |

To have some numbers of the problem we are dealing with, the previous table
shows the number of records cleaned by field for last runs at the moment of
writing this IP, except for tags, which we don't have accurate registries since
file saving was disabled.

The alternative is to upload TSV files to the Amazon Simple Storage Service
(S3), creating a new bucket or using a subfolder within `openverse-catalog`. The
benefit of using S3 buckets is that they have streaming capabilities and will
allow us to read the files in chunks later if necessary for performance. The
downside is that objects in S3 don't allow appending natviely, so it may require
to upload files with different part numbers or evaluate if the [multipart upload

> process][aws_mpu] or more easily, the `smart_open` package could serve us here.

Reviewer: Do you mind linking `smart_open` to the library here as well?

Author: This was supposed to be the first link! 😆  Fixed.
